docs/api/app.md

```diff
@@ -637,6 +637,8 @@ Returns `string` - The current application directory.
     * `~/Library/Application Support` on macOS
   * `userData` The directory for storing your app's configuration files, which by
     default it is the `appData` directory appended with your app's name.
+  * `browserData` The directory for storing browsing data, such as localStorage,
+    cookies, and disk cache. By default this points to `userData`.
```

Comment:
quite confused about this description, because it seems to me that localStorage and cookies should be in the "user data" directory.

Reply:
Chromium does not provide a way to set separate locations for localStorage and disk cache, so we currently put them all under `userData`. This PR means to provide a way to change the location of localStorage and disk cache, as there are apps willing to put them under cache locations (https://github.com/electron/electron/issues/1404).

In future I would like to add a new `browserCache` path to only set location for things like disk cache, but currently it is not practical as it requires a large patch to Chromium.

Reply:
Then I'm not sure I see the point of this change then? What then is stored in `userData`?

Reply:
For example in VS Code, user configurations are stored as JSON under `userData`, while there is no user information stored in localStorage or cookies.

Also apart from browser data, things like crash reports and app logs are currently stored under `userData`, the `requestSingleInstance` API also relies on `userData` to exchange data.

For apps don't rely on localStorage or cookies to store user data, it is necessary to separate `browserData` from `userData`.

Reply:
So it looks like this PR would group the following things under the new `browserData` directory:
- dictionaries
- browser prefs (`Local State`). not sure exactly what is stored in here, maybe things like devtools state?
- network service state. also not sure exactly what data is stored here. dns cache perhaps?
- ElectronBrowserContext's path, which is all the user profile data like localStorage, cookies, etc.
- some devtools-related temporary unix sockets

and things that remain in `userData` are:
- crash reports
- some caches?
- anything else?

Is this accurate? This is a weird grouping. I feel like `browserData` is the wrong name for this. At the very least this should be extensively documented, including exactly which data is stored in which directory.

Reply:
The idea behind the grouping is, users should be able to move all files generated by `BrowserContext` to a different directory. So dictionaries, browser prefs, devtools files, network caches belong to the same group with localStorage and cookies, as they are only generated when a `BrowserContext` is created.

I agree `browserData` is confusing, "browser" means lots of things in Electron. How do you think using `sessionData` as name instead? It should be clear that it is where `Session` code writes files to. If in future we add APIs to specify the path of `Session` to store data, it would point to `sessionData` by default.

  * `sessionData` The directory for storing data generated by `Session`, such as localStorage, cookies, disk cache, downloaded dictionaries, network state, devtools files. By default this points to `userData`.

Reply:
I don't think downloaded dictionaries are specific to a Session?

Reply:
Also, are we sure that no other "session-related" data is stored in userData? e.g. v8 code caches, IndexedDB data, service worker registration, etc. etc.

Reply:
Dictionaries are downloaded by Chromium automatically via the `ses.setSpellCheckerDictionaryDownloadURL` API, and all dictionary APIs belong to Session. In Chromium's context, the SpellcheckServiceFactory is attached to BrowserContext.

For other data, I can write tests to verify they are in the expected location.

Reply:
I'm somewhat worried about new session-related data being introduced in Chromium with no easy way for us to redirect where it's stored without patching Chromium. It also seems really easy to miss session-related data when implementing this. How can we be sure we caught all the cases?

Conceptually, though, the idea of a `sessionData` path doesn't seem totally unreasonable. It does raise the question from an API perspective of whether it should be possible to set session paths on a per-session basis.